### PR TITLE
[numpy] add numpy op fill_diagonal

### PR DESCRIPTION
**Pull request by BenjaminCHEN2016**
## Description ##
add numpy op fill_diagonal

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by haojin2** [comment]
@hzfan Could you double check on the ffi part?

File: src/api/operator/numpy/np_fill_diagonal_op.cc [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+.set_body([](runtime::MXNetArgs args, runtime::MXNetRetValue* ret) {
+  using namespace runtime;
+  const nnvm::Op* op = Op::Get("_npi_fill_diagonal");
+  nnvm::NodeAttrs attrs;
+
+  op::NumpyFillDiagonalParam param;
+  int num_inputs = 1;
+  NDArray* inputs[] = {args[0].operator mxnet::NDArray*()};
+
+  param.val = Tuple<double>(1, args[1].operator double());
```
Comment (hzfan) [inline src/api/operator/numpy/np_fill_diagonal_op.cc:39]: tuple of float is not supported for now. We may wait for #17866 .

File: tests/python/unittest/test_numpy_op.py [hunk 2]
```diff
@@ -6900,7 +6900,51 @@ def hybrid_forward(self, F, x, *args, **kwargs):
                         np_data[np_out] = -10
                         mx_data[mx_out] = -10
                         assert same(np_data, mx_data.asnumpy())
-                        
+
```
Comment (haojin2) [inline tests/python/unittest/test_numpy_op.py:6903]: one more blank line above

File: python/mxnet/symbol/numpy/_symbol.py [hunk 3]
```diff
@@ -6582,6 +6582,102 @@ def resize(a, new_shape):
...
 
 
+@set_module('mxnet.symbol.numpy')
+def fill_diagonal(a, val, wrap=False):
+    """
+    Fill the main diagonal of the given array of any dimensionality.
+    For an array `a` with ``a.ndim >= 2``, the diagonal is the list of
+    locations with indices ``a[i, ..., i]`` all identical. This function
+    modifies the input array in-place, it does not return a value.
+    Parameters
```
Comment (haojin2) [inline python/mxnet/symbol/numpy/_symbol.py:6592]: one more blank line above

File: python/mxnet/symbol/numpy/_symbol.py [hunk 4]
```diff
@@ -6582,6 +6582,102 @@ def resize(a, new_shape):
...
+    See also
+    --------
+    diag_indices, diag_indices_from
+    Notes
+    -----
+    .. versionadded:: 1.4.0
+    This functionality can be obtained via `diag_indices`, but internally
+    this version uses a much faster implementation that never constructs the
+    indices and uses simple slicing.
+    Examples
```
Comment (haojin2) [inline python/mxnet/symbol/numpy/_symbol.py:6612]: one more blank line above

File: python/mxnet/symbol/numpy/_symbol.py [hunk 5]
```diff
@@ -6582,6 +6582,102 @@ def resize(a, new_shape):
...
+      Value to be written on the diagonal, its type must be compatible with
+      that of the array a.
+    wrap : bool
+      For tall matrices in NumPy version up to 1.6.2, the
+      diagonal "wrapped" after N columns. You can have this behavior
+      with this option. This affects only tall matrices.
+    See also
+    --------
+    diag_indices, diag_indices_from
+    Notes
```
Comment (haojin2) [inline python/mxnet/symbol/numpy/_symbol.py:6606]: no need for `notes` and `see also`

File: python/mxnet/symbol/numpy/_symbol.py (hunk 4 above)
Comment (haojin2) [reply python/mxnet/symbol/numpy/_symbol.py:6612]: actually no need for `examples` for symbol

File: python/mxnet/ndarray/numpy/_op.py [hunk 6]
```diff
@@ -7477,6 +7477,94 @@ def resize(a, new_shape):
...
 
 
+@set_module('mxnet.ndarray.numpy')
+def fill_diagonal(a, val, wrap=False):
+    """
+    Fill the main diagonal of the given array of any dimensionality.
+    For an array `a` with ``a.ndim >= 2``, the diagonal is the list of
+    locations with indices ``a[i, ..., i]`` all identical. This function
+    modifies the input array in-place, it does not return a value.
+    Parameters
```
Comment (haojin2) [inline python/mxnet/ndarray/numpy/_op.py:7487]: one more blank line above

File: src/api/operator/numpy/np_fill_diagonal_op.cc (hunk 1 above)
Comment (haojin2) [reply src/api/operator/numpy/np_fill_diagonal_op.cc:39]: #17866 merged

File: src/operator/numpy/np_fill_diagonal_op-inl.h [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+    DMLC_DECLARE_FIELD(val)
+      .describe("Value to be written on the diagonal, "
+                "its type must be compatible with that of the array a.");
+    DMLC_DECLARE_FIELD(wrap)
+    .set_default(false)
+    .describe("The diagonal “wrapped” after N columns."
+              "You can have this behavior with this option. "
+              "This affects only tall matrices.");
+  }
+  void SetAttrDict(std::unordered_map<std::string, std::string>* dict) {
```
Comment (haojin2) [inline src/operator/numpy/np_fill_diagonal_op-inl.h:53]: one more blank line above

File: src/api/operator/numpy/np_fill_diagonal_op.cc (hunk 1 above)
Comment (hzfan) [reply src/api/operator/numpy/np_fill_diagonal_op.cc:39]: @BenjaminCHEN2016 #17866 has been merged. Let's use `Obj2Tuple` like https://github.com/apache/incubator-mxnet/pull/17866/files#diff-4c8cf94cd8bf4368b07f43c165402239R49.

File: python/mxnet/ndarray/numpy/_op.py [hunk 8]
```diff
@@ -7477,6 +7477,94 @@ def resize(a, new_shape):
...
+           [3, 0, 0]])
+    >>> np.fill_diagonal(np.flipud(a), [1,2,3])  # Vertical flip
+    >>> a
+    array([[0, 0, 3],
+           [0, 2, 0],
+           [1, 0, 0]])
+    Note that the order in which the diagonal is filled varies depending
+    on the flip function.
+    """
+    return _npi.fill_diagonal(a, val=val, wrap=wrap, out=a)
```
Comment (hzfan) [inline python/mxnet/ndarray/numpy/_op.py:7565]: Use `_api_internal`

**Comment by BenjaminCHEN2016** [comment]
@mxnet-bot run ci [windows-gpu]

**Comment by BenjaminCHEN2016** [comment]
@mxnet-bot run ci [windows-cpu]

**Comment by BenjaminCHEN2016** [comment]
@mxnet-bot run ci [windows-cpu, windows-gpu]

File: python/mxnet/numpy/multiarray.py [hunk 9]
```diff
@@ -9870,6 +9870,94 @@ def ones_like(a, dtype=None, order='C', ctx=None, out=None):
...
+           [3, 0, 0]])
+    >>> np.fill_diagonal(np.flipud(a), [1,2,3])  # Vertical flip
+    >>> a
+    array([[0, 0, 3],
+           [0, 2, 0],
+           [1, 0, 0]])
+    Note that the order in which the diagonal is filled varies depending
+    on the flip function.
+    """
+    return _mx_nd_np.fill_diagonal(a, val=val, wrap=wrap)
```
Comment (haojin2) [inline python/mxnet/numpy/multiarray.py:9958]: Actually there shouldn't be a return value for this function...This operator is an in-place operator

**Comment by BenjaminCHEN2016** [comment]
@mxnet-bot run ci [all]

**Comment by BenjaminCHEN2016** [comment]
@mxnet-bot run ci [all]

**Comment by BenjaminCHEN2016** [comment]
@mxnet-bot run ci [windows-gpu]

**Comment by BenjaminCHEN2016** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by BenjaminCHEN2016** [comment]
@mxnet-bot run ci [unix-gpu]

**Pull request merged**